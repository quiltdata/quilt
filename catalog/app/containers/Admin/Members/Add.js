import Button from 'material-ui/RaisedButton';
import TextField from 'material-ui/TextField';
import { red500 } from 'material-ui/styles/colors';
import PT from 'prop-types';
import React, { Fragment } from 'react';
import { FormattedMessage as FM, injectIntl } from 'react-intl';
import { setPropTypes } from 'recompose';
import { reduxForm, Field, SubmissionError } from 'redux-form/immutable';
import styled from 'styled-components';

import { push } from 'containers/Notifications/actions';
import Spinner from 'components/Spinner';
import { composeComponent } from 'utils/reactTools';
import * as validators from 'utils/validators';

import msg from './messages';

const FormField = composeComponent('Admin.Members.Add.FormField',
  setPropTypes({
    input: PT.object.isRequired,
    meta: PT.shape({
      submitting: PT.bool,
      submitFailed: PT.bool,
      error: PT.string,
    }).isRequired,
    errors: PT.objectOf(PT.string),
  }),
  ({ input, meta, errors = {}, ...props }) => (
    <TextField
      {...input}
      {...props}
      disabled={meta.submitting}
      errorText={meta.submitFailed ? errors[meta.error] || meta.error : undefined}
    />
  ));

const Form = styled.form`
  align-items: center;
  display: flex;
  flex-wrap: wrap;
`;

const FormError = styled.p`
  color: ${red500};
  flex-basis: 100%;
  font-family: Roboto !important;
  font-size: .8em;
`;

const FORM_ERRORS = {
  uniq: msg.addFormErrorUniq,
  username: msg.addFormErrorUsername,
};

export default composeComponent('Admin.Members.Add',
  injectIntl,
  setPropTypes({
    add: PT.func.isRequired,
    intl: PT.shape({
      formatMessage: PT.func.isRequired,
    }).isRequired,
  }),
  reduxForm({
    form: 'Admin.Members.Add',
    onSubmit: (values, _dispatch, { add }) =>
      new Promise((resolve, reject) => add(values.toJS(), { resolve, reject }))
        .catch((err) => {
          if (err.message === "Bad request. Maybe there's already a user with the username you provided?") {
            throw new SubmissionError({ _error: 'uniq' });
          }
          if (err.message === 'Please enter a valid email address.') {
            throw new SubmissionError({ email: 'email' });
          }
          if (/Username is not valid/.test(err.message)) {
            throw new SubmissionError({ username: 'username', _error: 'username' });
          }

          throw err;
        }),
    // eslint-disable-next-line object-curly-newline
    onSubmitSuccess: ({ name, email }, dispatch, { reset, intl: { formatMessage } }) => {
      reset();
      dispatch(push(formatMessage(msg.addSuccess, { name, email })));
    },
  }),
  ({
    handleSubmit,
    error,
    pristine,
    invalid,
    submitting,
    submitFailed,
    intl: { formatMessage },
  }) => (
    <Fragment>
      <h3><FM {...msg.addHeading} /></h3>
      <Form onSubmit={handleSubmit}>
        {error &&
          <FormError>
            {error in FORM_ERRORS ? formatMessage(FORM_ERRORS[error]) : error}
          </FormError>
        }
        <Field
          name="username"
          validate={[validators.required]}
          component={FormField}
          hintText={formatMessage(msg.addUsername)}
          errors={{
            required: formatMessage(msg.addUsernameRequired),
            username: formatMessage(msg.addUsernameInvalid),
          }}
          style={{ marginRight: '12px' }}
        />
        <Field
          name="email"
          validate={[validators.required]}
          component={FormField}
          hintText={formatMessage(msg.addEmail)}
          errors={{
            required: formatMessage(msg.addEmailRequired),
            email: formatMessage(msg.addEmailInvalid),
          }}
          style={{ marginRight: '12px' }}
        />
        <Button
          type="submit"
          label={formatMessage(msg.addSubmit)}
          disabled={submitting || pristine || (submitFailed && invalid)}
        />
        {submitting
          ? <Spinner style={{ fontSize: '2em', marginLeft: '12px', opacity: 0.5 }} />
          : null
        }
      </Form>
    </Fragment>
  ));
