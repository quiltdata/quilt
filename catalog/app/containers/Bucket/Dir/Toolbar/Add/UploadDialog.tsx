import { join } from 'path'

import * as React from 'react'
import * as M from '@material-ui/core'
import * as Lab from '@material-ui/lab'

import { useUploads } from 'containers/Bucket/PackageDialog/Uploads'
import * as FI from 'containers/Bucket/PackageDialog/FilesInput'
import type * as Toolbar from 'containers/Bucket/Toolbar'
import Log from 'utils/Logging'
import * as s3paths from 'utils/s3paths'

const useUploadDialogStyles = M.makeStyles((t) => ({
  drop: {
    height: t.spacing(60),
    marginBottom: t.spacing(1),
  },
  alert: {
    marginBottom: t.spacing(2),
  },
}))

const toFilesState = (added: FI.FilesState['added']): FI.FilesState => ({
  added,
  existing: {},
  deleted: {},
})

type UploadState =
  | { _tag: 'idle' }
  | { _tag: 'uploading' }
  | { _tag: 'success'; count: number }
  | { _tag: 'error'; error: Error }

interface UploadDialogProps {
  handle: Toolbar.DirHandle
  initial?: FI.FilesState['added']
  onClose: (reason?: 'upload-success') => void
}

export default function UploadDialog({
  handle,
  initial = {},
  onClose,
}: UploadDialogProps) {
  const [value, setValue] = React.useState<FI.FilesState['added']>(initial)
  const classes = useUploadDialogStyles()

  const [uploadState, setUploadState] = React.useState<UploadState>({ _tag: 'idle' })

  const submitting = uploadState._tag === 'uploading'
  const meta = React.useMemo(
    () => ({
      initial: toFilesState(initial),
      submitting,
    }),
    [initial, submitting],
  )

  const { progress, remove, removeByPrefix, reset, upload } = useUploads()
  const onFilesAction = React.useMemo(
    () =>
      FI.FilesAction.match({
        _: () => {},
        Revert: remove,
        RevertDir: removeByPrefix,
        Reset: reset,
      }),
    [remove, removeByPrefix, reset],
  )

  const onUpload = React.useCallback(async () => {
    setUploadState({ _tag: 'uploading' })

    const { local: files } = FI.groupAddedFiles(value)
    try {
      const uploadedEntries = await upload({
        files,
        bucket: handle.bucket,
        getCanonicalKey: (key) => s3paths.withoutPrefix('/', join(handle.path, key)),
        getMeta: () => null,
      })
      setUploadState({ _tag: 'success', count: Object.keys(uploadedEntries).length })
    } catch (e) {
      Log.error(e)
      const error = e instanceof Error ? e : new Error('Upload failed')
      setUploadState({ _tag: 'error', error })
    }
  }, [handle, upload, value])

  const handleClose = React.useCallback(() => {
    if (uploadState._tag === 'uploading') return
    if (uploadState._tag === 'success') {
      onClose('upload-success')
    } else {
      onClose()
    }
  }, [onClose, uploadState])

  const input = React.useMemo(
    () => ({
      value: toFilesState(value),
      // Trigger value update, because hashing finishes on the same objects
      onChange: (s: FI.FilesState) => setValue({ ...s.added }),
    }),
    [value],
  )

  const showUploadUI = uploadState._tag === 'idle' || uploadState._tag === 'uploading'

  return (
    <>
      <M.DialogContent>
        {uploadState._tag === 'success' && (
          <Lab.Alert severity="success" className={classes.alert}>
            Successfully uploaded {uploadState.count} file
            {uploadState.count !== 1 ? 's' : ''}!
          </Lab.Alert>
        )}

        {uploadState._tag === 'error' && (
          <Lab.Alert
            severity="error"
            className={classes.alert}
            onClose={() => setUploadState({ _tag: 'idle' })}
          >
            <Lab.AlertTitle>Upload Failed</Lab.AlertTitle>
            {uploadState.error.message}
          </Lab.Alert>
        )}

        {showUploadUI && (
          <FI.FilesInput
            className={classes.drop}
            input={input}
            meta={meta}
            onFilesAction={onFilesAction}
            title="Upload files"
            totalProgress={progress}
            validationErrors={null}
            noMeta
          />
        )}
      </M.DialogContent>
      <M.DialogActions>
        <M.Button
          color="primary"
          onClick={handleClose}
          disabled={uploadState._tag === 'uploading'}
        >
          {uploadState._tag === 'success' ? 'Close' : 'Cancel'}
        </M.Button>
        {uploadState._tag !== 'success' && (
          <M.Button
            color="primary"
            variant="contained"
            disabled={!Object.keys(value).length || uploadState._tag === 'uploading'}
            onClick={onUpload}
          >
            {uploadState._tag === 'uploading' ? 'Uploading...' : 'Upload'}
          </M.Button>
        )}
      </M.DialogActions>
    </>
  )
}
