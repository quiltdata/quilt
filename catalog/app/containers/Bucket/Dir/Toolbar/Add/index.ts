export { default as UploadDialog } from './UploadDialog'
export * as Context from './Context'
export { default as Options } from './Options'
