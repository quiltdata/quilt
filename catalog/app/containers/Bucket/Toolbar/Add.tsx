import * as React from 'react'

import { AddOutlined as IconAddOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

type ButtonProps = Omit<Buttons.WithPopoverProps, 'icon'>

export default function Button({ label = 'Add files', ...props }: ButtonProps) {
  return <Buttons.WithPopover icon={IconAddOutlined} label={label} {...props} />
}
