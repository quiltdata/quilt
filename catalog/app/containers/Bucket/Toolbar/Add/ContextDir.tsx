import { join } from 'path'

import invariant from 'invariant'
import * as React from 'react'
import * as M from '@material-ui/core'

import * as FileEditor from 'components/FileEditor'
import * as Dialogs from 'utils/Dialogs'
import Log from 'utils/Logging'
import * as s3paths from 'utils/s3paths'

import { useUploads } from '../../PackageDialog/Uploads'
import * as FI from '../../PackageDialog/FilesInput'
import type { DirHandle } from '../types'

interface AddDirActions {
  createFile: () => void
  openUploadDialog: () => void
}

const AddDirContext = React.createContext<AddDirActions | null>(null)

export function useAddDirActions(): AddDirActions {
  const context = React.useContext(AddDirContext)
  invariant(context, 'useAddDirActions must be used within AddDirProvider')
  return context
}

interface LocalEntry {
  file: FI.LocalFile
  path: string
}

const useUploadDialogStyles = M.makeStyles((t) => ({
  drop: {
    height: t.spacing(60),
    marginBottom: t.spacing(1),
  },
}))

interface UploadDialogProps {
  handle: DirHandle
  initial?: FI.FilesState['added']
  onClose: () => void
}

export function UploadDialog({ handle, initial = {}, onClose }: UploadDialogProps) {
  const [value, setValue] = React.useState<FI.FilesState['added']>(initial)
  const classes = useUploadDialogStyles()

  const [submitting, setSubmitting] = React.useState(false)
  const meta = React.useMemo(
    () => ({
      initial: { added: initial, existing: {}, deleted: {} },
      submitting,
    }),
    [initial, submitting],
  )

  const { progress, remove, removeByPrefix, reset, upload } = useUploads()
  const onFilesAction = React.useMemo(
    () =>
      FI.FilesAction.match({
        _: () => {},
        Revert: remove,
        RevertDir: removeByPrefix,
        Reset: reset,
      }),
    [remove, removeByPrefix, reset],
  )

  const onUpload = React.useCallback(async () => {
    setSubmitting(true)

    const files = Object.entries(value).map(
      ([path, file]) => ({ path, file }) as LocalEntry,
    )
    try {
      const uploadedEntries = await upload({
        files,
        bucket: handle.bucket,
        getCanonicalKey: (key) => s3paths.withoutPrefix('/', join(handle.path, key)),
        getMeta: () => null,
      })
      Log.log(uploadedEntries)
      // FIXME: show success
    } catch (e) {
      // FIXME: show error
      Log.error(e)
    }

    setSubmitting(false)
    onClose()
  }, [onClose, value, handle, upload])

  const input = React.useMemo(
    () => ({
      value: { added: initial, existing: {}, deleted: {} },
      onChange: ({ added }: FI.FilesState) => setValue(added),
    }),
    [initial, setValue],
  )

  return (
    <>
      <M.DialogContent>
        <FI.FilesInput
          className={classes.drop}
          input={input}
          meta={meta}
          onFilesAction={onFilesAction}
          title="Upload files"
          totalProgress={progress}
          validationErrors={null}
        />
      </M.DialogContent>
      <M.DialogActions>
        <M.Button color="primary" onClick={onClose}>
          Cancel
        </M.Button>
        <M.Button
          color="primary"
          variant="contained"
          disabled={!Object.keys(value.added).length || meta.submitting}
          onClick={onUpload}
        >
          Upload
        </M.Button>
      </M.DialogActions>
    </>
  )
}

interface AddDirProviderProps {
  children: React.ReactNode
  handle: DirHandle
}

export function AddDirProvider({ children, handle }: AddDirProviderProps) {
  const dialogs = Dialogs.use()
  const prompt = FileEditor.useCreateFileInBucket(handle.bucket, handle.path)

  const createFile = React.useCallback(() => {
    prompt.open()
  }, [prompt])

  const openUploadDialog = React.useCallback(() => {
    dialogs.open(({ close }) => <UploadDialog handle={handle} onClose={close} />)
  }, [dialogs, handle])

  const actions = React.useMemo(
    (): AddDirActions => ({
      createFile,
      openUploadDialog,
    }),
    [createFile, openUploadDialog],
  )

  return (
    <AddDirContext.Provider value={actions}>
      {children}

      {prompt.render()}

      {dialogs.render({ fullWidth: true, maxWidth: 'sm' })}
    </AddDirContext.Provider>
  )
}

export { AddDirProvider as Provider }
