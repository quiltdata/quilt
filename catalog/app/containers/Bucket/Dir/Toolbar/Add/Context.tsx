import invariant from 'invariant'
import * as React from 'react'

import * as FileEditor from 'components/FileEditor'
import type * as Toolbar from 'containers/Bucket/Toolbar'
import * as Dialogs from 'utils/Dialogs'

import UploadDialog from './UploadDialog'

interface AddState {
  createFile: () => void
  openUploadDialog: () => void
}

const Context = React.createContext<AddState | null>(null)

function useContext(): AddState {
  const context = React.useContext(Context)
  invariant(context, 'useContext must be used within AddDirProvider')
  return context
}

export const use = useContext

interface AddProviderProps {
  children: React.ReactNode
  handle: Toolbar.DirHandle
}

export function AddProvider({ children, handle }: AddProviderProps) {
  const dialogs = Dialogs.use()
  const prompt = FileEditor.useCreateFileInBucket(handle.bucket, handle.path)

  const createFile = React.useCallback(() => {
    prompt.open()
  }, [prompt])

  const openUploadDialog = React.useCallback(() => {
    dialogs.open(({ close }) => <UploadDialog handle={handle} onClose={close} />)
  }, [dialogs, handle])

  const actions = React.useMemo(
    (): AddState => ({
      createFile,
      openUploadDialog,
    }),
    [createFile, openUploadDialog],
  )

  return (
    <Context.Provider value={actions}>
      {children}

      {prompt.render()}

      {dialogs.render({ fullWidth: true, maxWidth: 'sm' })}
    </Context.Provider>
  )
}

export { AddProvider as Provider }
