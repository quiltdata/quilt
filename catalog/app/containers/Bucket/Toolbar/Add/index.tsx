import * as React from 'react'

import { AddOutlined as IconAddOutlined } from '@material-ui/icons'

import * as Buttons from 'components/Buttons'

import type { DirHandle } from '../types'

import * as Context from './ContextDir'

export { default as BucketDirOptions } from './BucketDirOptions'
export { default as UploadDialog } from './UploadDialog'

interface ButtonProps {
  children: NonNullable<React.ReactNode>
  className?: string
  handle: DirHandle
  label?: string
}

export function Button({ children, handle, label = 'Add files', ...props }: ButtonProps) {
  return (
    <Context.Provider handle={handle}>
      <Buttons.WithPopover icon={IconAddOutlined} label={label} {...props}>
        {children}
      </Buttons.WithPopover>
    </Context.Provider>
  )
}
