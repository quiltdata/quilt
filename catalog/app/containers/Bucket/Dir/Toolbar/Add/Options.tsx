import * as React from 'react'
import * as M from '@material-ui/core'
import * as Icons from '@material-ui/icons'

import * as Context from './Context'

const LIST_ITEM_TYPOGRAPHY_PROPS = { noWrap: true } as const

interface MenuItemProps {
  icon: React.ReactElement
  primary: string
  onClick: () => void
}

function MenuItem({ icon, primary, onClick }: MenuItemProps) {
  return (
    <M.ListItem button onClick={onClick}>
      <M.ListItemIcon>{icon}</M.ListItemIcon>
      <M.ListItemText
        primary={primary}
        primaryTypographyProps={LIST_ITEM_TYPOGRAPHY_PROPS}
      />
    </M.ListItem>
  )
}

export default function AddOptions() {
  const { createFile, openUploadDialog } = Context.use()
  return (
    <M.List dense>
      <MenuItem
        icon={<Icons.CreateOutlined />}
        primary="Create text file"
        onClick={createFile}
      />
      <MenuItem
        icon={<Icons.PublishOutlined />}
        primary="Upload files"
        onClick={openUploadDialog}
      />
    </M.List>
  )
}
