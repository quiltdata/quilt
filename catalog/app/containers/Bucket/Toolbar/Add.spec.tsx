import * as React from 'react'
import renderer from 'react-test-renderer'

import Add from './Add'

jest.mock('@material-ui/core', () => ({
  ...jest.requireActual('@material-ui/core'),
  IconButton: ({ className, children }: any) => <button {...{ className, children }} />,
  Button: ({ className, children }: any) => <button {...{ className, children }} />,
}))

describe('containers/Bucket/Toolbar/Add', () => {
  it('should render with default label', () => {
    const tree = renderer
      .create(<Add className="custom-class">Hello, Popover!</Add>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render with custom label', () => {
    const tree = renderer
      .create(<Add label="Custom Add Label">Hello, Popover!</Add>)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
