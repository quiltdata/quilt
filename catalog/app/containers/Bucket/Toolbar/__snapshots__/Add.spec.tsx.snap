// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/Bucket/Toolbar/Add should render with custom label 1`] = `
<div
  className="makeStyles-root-1"
>
  <button>
    Custom Add Label
  </button>
  <div
    aria-hidden={true}
    className="MuiBackdrop-root makeStyles-backdrop-3"
    onClick={[Function]}
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  />
</div>
`;

exports[`containers/Bucket/Toolbar/Add should render with default label 1`] = `
<div
  className="makeStyles-root-1"
>
  <button
    className="custom-class"
  >
    Add files
  </button>
  <div
    aria-hidden={true}
    className="MuiBackdrop-root makeStyles-backdrop-3"
    onClick={[Function]}
    style={
      {
        "opacity": 0,
        "visibility": "hidden",
      }
    }
  />
</div>
`;
